### ISSUE TITLE
C# 8 Nullable - [NotNullWhen(bool)] does not work with Task<bool>

### ISSUE THREAD

**Issue by MarcusKaseder** [dotnet/roslyn#45228]
**Version Used**: 
VS-Enterprise 2019 16.5.4

**Steps to Reproduce**:

`        private async Task NotNullTestAsync(bool fakeSyncParameter, bool fakeAsyncParameter)
        {
            // SYNC
            var syncValue = fakeSyncParameter ? null : (int?)0;

            if (!this.NotNullWhenTrueSync(syncValue))
            {
                return;
            }

            // Works...
            Debug.WriteLine(syncValue.Value);

            // ASYNC
            var asyncValue = fakeAsyncParameter ? null : (int?)0;
            var asyncResult = await this.NotNullWhenTrueAsync(asyncValue).ConfigureAwait(false);

            if (!asyncResult)
            {
                return;
            }

            // Nullable value type may be null "error"
            Debug.WriteLine(asyncValue.Value);
        }

        private bool NotNullWhenTrueSync([NotNullWhen(true)]int? value)
        {
            return value.HasValue;
        }

        private async Task<bool> NotNullWhenTrueAsync([NotNullWhen(true)]int? value)
        {
            return await Task.FromResult(value.HasValue).ConfigureAwait(false);
        }`

**Expected Behavior**:
Async works as the sync example

**Actual Behavior**:
Task<bool> is not taken into acccount for [NotNullWhen(true)] Attribute

**Comment by RikkiGibson** [comment]
@jcouv do you have any sense of how easy/hard this would be to implement? or whether it makes sense for us to do this for 'async' methods?

**Comment by jcouv** [comment]
I think we could probably fix `await` to create a proper split state in a scenario like `if (await MethodAsync(...)) ...`. That's probably not very difficult.

But I don't think we'll be able to fix this scenario as presented, because it has multiple complications:
1. the use of `.ConfigureAwait(...)` will collapse the split state of evaluating `this.NotNullWhenTrueAsync(asyncValue)` (this seems fixable with special knowledge of this method)
2. the use of a local variable to store the boolean result means we'll lose track (that one is a wicked problem)

([repro on sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHABgAIcCBWAblIuoIDoARNCAcwDsB7WGgDGURmSo0OXPoJgiorAML8AJnACCvCBgCeUNGNIBiXgFcMGCMAxxKcbTbikmAZmp5KSygG9SlAMDKAAcENAA3CHhqAA5qADZKADl+GCSLDAAVOFgNKF1eYQAKYH5+DEoAMwgAazgAZQLhAAUIBAgwOHhkSlLyqtrNfMLW9s7ugEp/IMC/EhmFygB6Jcp6gE0kpWnFwMiESmHhADUdMzsAXgG6xpG2jq7ESgB+SnNLShBKIrReGGeJuRxDtdpQ0JVvgBCGAAC0MrBSaQyAHUYQ5MghzrdikdThhzhMpvNQQE5iSFjgAOzicmUAC+LmJJJWlGR/AQNUUrFYIN27DgwDM3FYyLC8AAMr84EVcWc4Kw8QTgUzQSyNBstrzFvtYorLtchk1Rg9ui83hlPt9fv9ATTyTqcDEAEo5CwwShXHAATkosPhiPSllR6MxhsKRUdeomyn4vEqaG4ZgQmgA7hA0DAitUMFA4BNlbTwVDHS6oG6ibSybTAlS7SSGSQtQsWYGrE5KJF8XYYLpgnZILpenZ3hUAESIBDs0dNmb8wXC0UZuCS3jSyNyhVy/Mzhsz0IRKJ2PoVAMotG8DFYppFADap6D56KMFDEwAutbXp2CTOqySqR2NwACWgPU6xmXcVSCfdImiR0EgAHmPAA+ZJUlbYML1DPJrzvNCzwcJ8X3fP5Py3H8ZyCf9vQSVgADFJzAUs3SKL95WAqAoxjOMEyTVN00zbNc23SDAgbOkgA==))

**Comment by RikkiGibson** [comment]
Hmm. it also feels like there is more complexity to analyzing the implementation of `NotNullWhenTrueAsync`. Basically we would also want to thread the conditional state through the Task.FromResult call.

**Comment by jcouv** [comment]
@RikkiGibson I think the present issue is about the *caller* of  `Task<bool> NotNullWhenTrueAsync([NotNullWhen(true)]int? value)`, not the implementation of that method. So I don't think there is anything to do about `Task.FromResult`.

**Comment by MarcusKaseder** [comment]
> @RikkiGibson I think the present issue is about the _caller_ of `Task<bool> NotNullWhenTrueAsync([NotNullWhen(true)]int? value)`, not the implementation of that method. So I don't think there is anything to do about `Task.FromResult`.

Yeah, that's right. I've added `Task.FromResult` just for demonstration purpose.

Thanks in advance. You're doing a great job and I love the new nullable feature :)

**Comment by johnknoop** [comment]
If this is indeed considered a bug (as the label suggests), does that mean this will be addressed any time soon?

**Comment by Arturace** [comment]
This looks like a rather complicated issue, but with no impact on the runtime and slight impact on the developer experience.
So, I understand that there is no urgency, but it has been 2 years. Is there any update on how this should be handled? Maybe another more important issue that also impacts Task boxing and unboxing for code analysis attributes?

**Comment by CyrusNajmabadi** [comment]
> Is there any update on how this should be handled? 

Right now this isn't supported.  So if you run into issues, just suppress them in them in teh code.  This is something we're interesting in a future release, along with stuff like making it so that `Task<string>` can be implicitly converted to things like `Task<string?>`.  But it won't be for a while as this is low on our list.
